### PR TITLE
Refactoring recipe execution

### PR DESCRIPTION
**Pull request by sebastienros**
Goal is to make it simpler to understand and maintain.

- Removed IRecipeExecutionStep and IRecipeExecutionStep in favor of IRecipeHandler.
- All IRecipeHandler implementations are called and are responsible for executing the steps they want
- Added environment object to pass parameters, and `parameters()` method provider
- Removed RecipeExecutionStep, RecipeStepDescriptor and RecipeContext in favor of a single RecipeExecutionContext.
- Added more fields in RecipeResultIndex to query current status of recipe execution
- Removed `RecipeExecutionStep` base class as it was just requiring a localizer and a logger without actually using it

### CODE REVIEW

File: src/Orchard.Cms.Web/Modules/Orchard.Recipes/RecipeSteps/RecipesStep.cs [hunk 1]
```diff
@@ -35,49 +31,38 @@ public class RecipesStep : RecipeExecutionStep
...
+
+            var step = context.Step.ToObject<InternalStep>();
             var recipesDictionary = new Dictionary<string, IDictionary<string, RecipeDescriptor>>();
 
             foreach (var recipe in step.Values)
             {
-                Logger.LogInformation("Executing recipe '{0}' in extension '{1}'.", recipe.Name, recipe.ExecutionId);
-
-                try
+                var recipes = recipesDictionary.ContainsKey(recipe.ExecutionId) ? recipesDictionary[recipe.ExecutionId] : default(IDictionary<string, RecipeDescriptor>);
```
Comment (Jetski5822) [inline src/Orchard.Cms.Web/Modules/Orchard.Recipes/RecipeSteps/RecipesStep.cs]: why default IDictionary?

File: src/Orchard.Recipes.Abstractions/Services/IRecipeExecutor.cs [hunk 2]
```diff
@@ -1,10 +1,10 @@
...
-using System.Threading.Tasks;
+﻿using System.Threading.Tasks;
+using Orchard.Recipes.Models;
 
 namespace Orchard.Recipes.Services
 {
     public interface IRecipeExecutor
     {
-        Task<string> ExecuteAsync(string executionId, RecipeDescriptor recipeDescriptor);
+        Task<string> ExecuteAsync(string executionId, RecipeDescriptor recipeDescriptor, object environment);
```
Comment (Jetski5822) [inline src/Orchard.Recipes.Abstractions/Services/IRecipeExecutor.cs]: not sure how I feel about passing just a random object through.... it feel like it allows the user to place fast and loose with the execute method.

File: src/Orchard.Cms.Web/Modules/Orchard.Recipes/RecipeSteps/RecipesStep.cs (hunk 1 above)
Comment (sebastienros) [inline src/Orchard.Cms.Web/Modules/Orchard.Recipes/RecipeSteps/RecipesStep.cs]: I agree, this is some really stupid code ;) oh, wait, it's yours 😄 I just moved it. I will fix.

File: src/Orchard.Recipes.Abstractions/Services/IRecipeExecutor.cs (hunk 2 above)
Comment (sebastienros) [inline src/Orchard.Recipes.Abstractions/Services/IRecipeExecutor.cs]: The goal of this object is to contain any value that a recipe might require. Recipes will be able to expose a `parameters` object describing the list of values they require, so we can build some UI and gather the value for it. What type of object would be better? I could have chosen a Dictionary, a dynamic, a JObject, in the end I decided on object. Make sense?

**Pull request merged**